src/doc/trpl/choosing-guarantees.md

```diff
@@ -0,0 +1,417 @@
+needs a copy of the reference. `&mut T` is not `Copy` because mutable aliases cannot be shared, and
+`&mut T` "owns" the data it points to somewhat since it can mutate.
+
+Practically speaking, a type can be `Copy` if a copy of its stack representation doesn't violate
+memory safety.
+
+[^2]: By "stack representation" I mean the data on the stack when a value of this type is held on the stack. For example, a `Vec<T>` has a stack representation of a pointer and two integers (length, capacity). While there is more data behind the indirection of the pointer, it is not part of the stack-held portion of the `Vec`. Looking at this a different way, a type is `Copy` if a `memcopy` of the data copies all the data owned by it.
+
+## `&T` and `&mut T` 
+
+These are immutable and mutable references respectively. They follow the "read-write lock" pattern,
+such that one may either have only one mutable reference to some data, or any number of immutable
+ones, but not both. This guarantee is enforced at compile time, and has no visible cost at runtime.
+In most cases such pointers suffice for sharing cheap references between sections of code.
+
+These pointers cannot be copied in such a way that they outlive the lifetime associated with them.
+
+## `*const T` and `*mut T`
+
+These are C-like raw pointers with no lifetime or ownership attached to them. They just point to
+some location in memory with no other restrictions. The only guarantee that these provide is that
+they cannot be dereferenced except in code marked `unsafe`.
+
+These are useful when building safe, low cost abstractions like `Vec<T>`, but should be avoided in
+safe code.
+
+## `Rc<T>`
+
+This is the first wrapper we will cover that has a runtime cost.
+
+
+[`Rc<T>`][rc] is a reference counted pointer. In other words, this lets us have multiple "owning"
+pointers to the same data, and the data will be freed (destructors will be run) when all pointers
+are out of scope.
+
+Internally, it contains a shared "reference count", which is incremented each time the `Rc` is
+cloned, and decremented each time one of the `Rc`s goes out of scope. The main responsibility of
+`Rc<T>` is to ensure that destructors are called for shared data.
+
+The internal data here is immutable, and if a cycle of references is created, the data will be
+leaked. If we want data that doesn't leak when there are cycles, we need a garbage collector.
+
+// Should I link to rust-gc and rajan-bacon-cc here?
+
+
+#### Guarantees
+
+The main guarantee provided here is that the data will not be destroyed until all references to it
+are out of scope.
+
+This should be used when you wish to dynamically allocate and share some data (read-only) between
+various portions of your program, where it is not certain which portion will finish using the
+pointer last. It's a viable alternative to `&T` when `&T` is either impossible to statically check
```

Comment:
This sentence is secretly about destructors, so it's really about use/ownership than referencing. But "owning the pointer" has other connotations and `Rc` is "shared ownership", so I don't want to use that term here.